### PR TITLE
fix: various fixes in message keystore

### PR DESCRIPTION
**Pull request by D4ryl00**
- use the first/last index value when computing push group references
- fix a bug that precomputes 100 keys each time we consume one key in the cache. Now, we precompute 100 keys only when we register a new device secret and we precompute only one key when we consume a key.
- add comments on functions

### Other

Set `TestReactivateAccountGroup` as a flappy test (the instability has no link with this PR).

### CODE REVIEW

File: pkg/cryptoutil/keystore_message_utils_test.go [hunk 1]
```diff
@@ -520,3 +522,78 @@ func TestMessageKeyHolderSubscription(t *testing.T) {
...
+
+	mkh, cleanup := cryptoutil.NewInMemMessageKeystore(logger)
+	defer cleanup()
+
+	// test with the ds counter
+	updateAndTestPushGroupReferences(ctx, mkh, devicePK, ds.Counter, g, t)
+
+	// do the same test with a new device secret counter
+	// so we can test if old references are deleted
+	updateAndTestPushGroupReferences(ctx, mkh, devicePK, ds.Counter+10, g, t)
```
Comment (glouvigny) [inline pkg/cryptoutil/keystore_message_utils_test.go:554]: What is the expected behaviour with an out of bounds value such as `ds.Counter + precomputePushRefsCount + 1`? Can you test it?

File: pkg/cryptoutil/keystore_message.go [hunk 2]
```diff
@@ -558,6 +625,9 @@ func (m *MessageKeystore) putDeviceChainKey(ctx context.Context, groupPK, device
 	return nil
 }
 
+// SealEnvelope seals the given payload and returns the sealed envelope.
+// It retrieves the device chain key from the keystore to seal the envelope.
+// It also updates the device secret and stores the next message key in the cache.
```
Comment (glouvigny) [inline pkg/cryptoutil/keystore_message.go:630]:
```suggestion
// SealEnvelope encrypts the given payload and returns it as an envelope to be
// published on the group's store. It retrieves the device's chain key from the
// keystore to encrypt the payload using symmetric encryption.
// The payload is signed using the device's long term private key for the
// target group.
// It also updates the device secret and stores the next message key
// in the cache.

```

File: pkg/cryptoutil/keystore_message.go [hunk 3]
```diff
@@ -671,6 +728,9 @@ func NewInMemMessageKeystore(logger *zap.Logger) (*MessageKeystore, func()) {
 	return NewMessageKeystore(ds, logger), func() { _ = ds.Close() }
 }
 
+// OpenOutOfStoreMessage opens the given OutOfStoreMessage and returns the decrypted payload.
+// The signature is verified against the given devicePK.
+// It derive the next message key and stores it in the cache, but it doesn't update the device secret.
```
Comment (glouvigny) [inline pkg/cryptoutil/keystore_message.go:733]:
```suggestion
// It derives the next message key and stores it in the cache, but it doesn't update the device secret.
```

File: deactivate_test.go [hunk 4]
```diff
@@ -17,6 +17,8 @@ import (
 )
 
 func TestReactivateAccountGroup(t *testing.T) {
+	testutil.FilterStability(t, testutil.Flappy)
```
Comment (glouvigny) [inline deactivate_test.go:20]: Why is this test now marked as flappy?

**Review by glouvigny — COMMENTED** [review]
I found "in the cache namespace" sounds weird when you are not mentioning the datastore

Other comments are linked to the code

File: pkg/cryptoutil/keystore_message_utils_test.go (hunk 1 above)
Comment (D4ryl00) [reply pkg/cryptoutil/keystore_message_utils_test.go:554]: yes, I already did that at the end of this test, lines 584 and 593.

**Review by glouvigny — APPROVED** [review]

**Pull request merged**